src/sentry/middleware/ratelimit.py

```diff
@@ -15,80 +15,98 @@
     get_rate_limit_value,
 )
 from sentry.ratelimits.config import ENFORCE_CONCURRENT_RATE_LIMITS
-from sentry.types.ratelimit import RateLimitCategory, RateLimitMeta, RateLimitType
+from sentry.types.ratelimit import RateLimitCategory, RateLimitType
 
 DEFAULT_ERROR_MESSAGE = (
     "You are attempting to use this endpoint too frequently. Limit is "
     "{limit} requests in {window} seconds"
 )
 
 
-class RatelimitMiddleware(MiddlewareMixin):
-    """Middleware that applies a rate limit to every endpoint."""
+class RatelimitMiddleware:
+    """Middleware that applies a rate limit to every endpoint.
+    See: https://docs.djangoproject.com/en/4.0/topics/http/middleware/#writing-your-own-middleware
+    """
 
-    def process_view(self, request: Request, view_func, view_args, view_kwargs) -> Response | None:
-        """Check if the endpoint call will violate."""
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request: Request) -> Response:
+        response, rate_limit_data = self.process_request(request)
+        # Hit the endpoint
+        request.rate_limit_metadata = rate_limit_data.get("rate_limit_metadata")
+        if not response:
+            response = self.get_response(request)
+
+        # Process the response
+        self.add_headers(response, request.rate_limit_metadata)
+        if rate_limit_data.get("rate_limit_key") and rate_limit_data.get("rate_limit_uid"):
+            finish_request(rate_limit_data["rate_limit_key"], rate_limit_data["rate_limit_uid"])
+        return response
+
+    def process_request(self, request: Request):
+        rate_limit_data = {}
```

Comment:
Since there's an effort to type the sentry codebase right now, let's be good citizens and make this function return a dataclass as opposed to a dict[Any, Any]. That way we can be more explicit about what this function is returning and the typechecker can do more work for us.